internal/pkg/service/stream/encryption/encryption.go

```diff
@@ -0,0 +1,72 @@
+package encryption
+
+import (
+	"context"
+	"crypto/rand"
+	"math/big"
+
+	"github.com/keboola/go-cloud-encrypt/pkg/cloudencrypt"
+)
+
+const (
+	ProviderNative = Provider("native")
+	ProviderGCP    = Provider("gcp")
+	ProviderAWS    = Provider("aws")
+	ProviderAzure  = Provider("azure")
+)
+
+type Provider string
+
+func NewEncryptor(ctx context.Context, config Config) (cloudencrypt.Encryptor, error) {
+	var encryptor cloudencrypt.Encryptor
+	var err error
+
+	switch config.Provider {
+	case ProviderNative:
+		encryptor, err = cloudencrypt.NewNativeEncryptor([]byte(config.Native.SecretKey))
+		if err != nil {
+			return nil, err
+		}
+	case ProviderGCP:
+		encryptor, err = cloudencrypt.NewGCPEncryptor(ctx, config.GCP.KMSKeyID)
+		if err != nil {
+			return nil, err
+		}
+	case ProviderAWS:
+		encryptor, err = cloudencrypt.NewAWSEncryptor(ctx, config.AWS.Region, config.AWS.KMSKeyID)
+		if err != nil {
+			return nil, err
+		}
+	case ProviderAzure:
+		encryptor, err = cloudencrypt.NewAzureEncryptor(ctx, config.Azure.KeyVaultURL, config.Azure.KeyName)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if config.Provider != ProviderNative {
+		encryptor, err = cloudencrypt.NewDualEncryptor(ctx, encryptor)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
```suggestion
        if config.Provider == ProviderNative {
                encryptor, err = cloudencrypt.NewNativeEncryptor([]byte(config.Native.SecretKey))
		if err != nil {
			return nil, err
		}
		
		return encryptor, nil
        }
        
	switch config.Provider {
	case ProviderGCP:
		encryptor, err = cloudencrypt.NewGCPEncryptor(ctx, config.GCP.KMSKeyID)
		if err != nil {
			return nil, err
		}
	case ProviderAWS:
		encryptor, err = cloudencrypt.NewAWSEncryptor(ctx, config.AWS.Region, config.AWS.KMSKeyID)
		if err != nil {
			return nil, err
		}
	case ProviderAzure:
		encryptor, err = cloudencrypt.NewAzureEncryptor(ctx, config.Azure.KeyVaultURL, config.Azure.KeyName)
		if err != nil {
			return nil, err
		}
	}

	encryptor, err = cloudencrypt.NewDualEncryptor(ctx, encryptor)
	if err != nil {
		return nil, err
        }
```
Basically the code want's to say that we do not want to create dual encryptor from native encryptor, so it goes on the top as first decision. When we have cloud specific vendor we do the switch and wrap it up with DualEncryptor always. 
We get rid of the if statement at the end which would someone refactor by putting the code into switch statement which was my first iteration.
After while I realised we can do simple if statement to go with fast path decision and rest of construct remains unchanged without any other decisions.
So we do not duplicate code and cut out the statement from `switch` into early `if` check
For me this is more readable that we have just one if statement at start and do not evaluate if over each `switch` case.

Reply:
I considered that but I think the current way is better because this way we can easily wrap the encryptor by logging encryptor for instance.

Reply:
You can do it also with if statement at the top using `defer` function. For me it is logically worse to read the code above because it has some if conditions at the end that we do not want to wrap dual encryptor to basic one, but why do I need to know such detail ? Simply put the if statement on top of that and you do not need if statement at the end or even you can use `defer` for wrapping up the `DualEncryptor`

Reply:
As you wish. https://github.com/keboola/keboola-as-code/commit/16bfb39a0d06011d971624f2aad6727714b99b46

Reply:
Still not the best but better. The `return` statement is quite hidden.